vsintegration/tests/FSharp.Editor.Tests/Refactors/AddExplicitReturnType.fs

```diff
@@ -0,0 +1,199 @@
+﻿module FSharp.Editor.Tests.Refactors.AddExplicitReturnType
+
+open Microsoft.VisualStudio.FSharp.Editor
+open Xunit
+open System
+open System.Collections.Immutable
+open System.Text.RegularExpressions
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.CodeFixes
+open Microsoft.CodeAnalysis.Text
+open Microsoft.VisualStudio.FSharp.Editor
+open Microsoft.VisualStudio.FSharp.Editor.CancellableTasks
+
+open FSharp.Compiler.Diagnostics
+open FSharp.Editor.Tests.Helpers
+
+open Microsoft.CodeAnalysis.CodeRefactorings
+open NUnit.Framework
+open Microsoft.CodeAnalysis.CodeActions
+open System.Collections.Generic
+open Microsoft.VisualStudio.LanguageServices
+open FSharp.Editor.Tests.Refactors.RefactorTestFramework
+open Microsoft.Build.Utilities
+open System.Threading
+open FSharp.Test.ReflectionHelper
+open Microsoft.Build.Utilities
+open FSharp.Test.ProjectGeneration.ProjectOperations
+open FSharp.Compiler.Symbols
+open Xunit
+open System.Runtime.InteropServices
+
+[<Theory>]
+[<InlineData(":int")>]
+[<InlineData(" :int")>]
+[<InlineData(" : int")>]
+[<InlineData(" :    int")>]
+let ``Refactor should not trigger`` (shouldNotTrigger: string) =
+    let symbolName = "sum"
+
+    let code =
+        $"""
+            let sum a b {shouldNotTrigger}= a + b
+            """
+
+    use context = TestContext.CreateWithCode code
+
+    let spanStart = code.IndexOf symbolName
+
+    let actions =
+        tryGetRefactoringActions code spanStart context (new AddExplicitReturnType())
+
+    do Assert.Empty(actions)
+
+[<Fact>]
+let ``Correctly infer int as explicit return type`` () =
+    let symbolName = "sum"
+
+    let code =
+        """
+        let sum a b = a + b
+        """
+
+    use context = TestContext.CreateWithCode code
+
+    let spanStart = code.IndexOf symbolName
+
+    let newDoc = tryRefactor code spanStart context (new AddExplicitReturnType())
+
+    let expectedCode =
+        $"""
+        let sum a b :int= a + b
+        """
+
+    let resultText = newDoc.GetTextAsync context.CT |> GetTaskResult
+    Assert.Equal(expectedCode, resultText.ToString())
+
+[<Fact>]
+let ``Handle Parantheses on the arguments`` () =
+    let symbolName = "sum"
+
+    let code =
+        """
+        let sum (a:float) (b:float) = a + b
+        """
+
+    use context = TestContext.CreateWithCode code
+
+    let spanStart = code.IndexOf symbolName
+
+    let newDoc = tryRefactor code spanStart context (new AddExplicitReturnType())
+
+    let expectedCode =
+        """
+        let sum (a:float) (b:float) :float= a + b
+        """
+
+    let resultText = newDoc.GetTextAsync context.CT |> GetTaskResult
+    Assert.Equal(expectedCode, resultText.ToString())
+
+[<Fact>]
+let ``Infer on rec method`` () =
+    let symbolName = "fib"
```

Comment:
@psfinaki This case and the member function case still give me "UndefinedName\n  (0, <fun:failingCase@3047>, factor0, <fun:suggestNamesAndTypes@3027>)" Exceptions. Any Idea why? The code is correct in FSI.

Reply:
@SebastianAtWork looks like it's just a spacing problem here. In the test, there is no indent on the 2nd line:
```fsharp
let rec fib n =
if n < 2 then 1 else fib (n - 1) + fib (n - 2)
```
The code doesn't fully compile, but with all the parser recovery, the compiler still infers the function parameter as a generic one:
![image](https://github.com/dotnet/fsharp/assets/5451366/026ef5b8-88b3-44ac-80f9-1f5bd5d4de7f)
And hence the code reasonably throws when you try to access the named type.

If I add the tab, the test passes for me:
```fsharp
let rec fib n =
    if n < 2 then 1 else fib (n - 1) + fib (n - 2)
```

Maybe you have correct spacing in the FSI...

Reply:
Same applies to the other test.

Actually I'd recommend pushing all the code to the wall: it will be just more compact and natural. Like:
```fsharp
[<Fact>]
let ``Infer on member function`` () =
    let symbolName = "SomeMethod"

    let code = $"""
type SomeType(factor0: int) =
    let factor = factor0
    member this.SomeMethod(a, b, c) = (a + b + c) * factor
"""

```

Reply:
Good point, Done